File: services/python/shared/middleware.py
import json
import time
import uuid
import asyncio
import logging
from collections import defaultdict
from dataclasses import dataclass, field
from enum import Enum
from functools import wraps
from typing import Callable, Optional

from fastapi import Request, Response
from starlette.middleware.base import BaseHTTPMiddleware


class TracingMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next):
        corr_id = request.headers.get("X-Correlation-ID", f"corr-{uuid.uuid4().hex[:12]}")
        trace_id = request.headers.get("X-Trace-ID", f"trace-{uuid.uuid4().hex[:16]}")
        span_id = f"span-{uuid.uuid4().hex[:12]}"

        request.state.correlation_id = corr_id
        request.state.trace_id = trace_id
        request.state.span_id = span_id

        response = await call_next(request)
        response.headers["X-Correlation-ID"] = corr_id
        response.headers["X-Trace-ID"] = trace_id
        response.headers["X-Span-ID"] = span_id
        return response


class PrometheusMetrics:
    def __init__(self, service_name: str):
        self.service_name = service_name
        self.request_count: dict[str, int] = defaultdict(int)
        self.error_count: dict[str, int] = defaultdict(int)
        self.latency_sum: dict[str, float] = defaultdict(float)
        self.latency_count: dict[str, int] = defaultdict(int)
        self.buckets = [5, 10, 25, 50, 100, 250, 500, 1000, 2500, 5000, 10000]
        self.bucket_counts: dict[str, list[int]] = {}

    def record(self, method: str, path: str, status_code: int, duration_ms: float):
        key = f"{method}_{path}"
        self.request_count[key] += 1
        if status_code >= 400:
            self.error_count[key] += 1
        self.latency_sum[key] += duration_ms
        self.latency_count[key] += 1

        if key not in self.bucket_counts:
            self.bucket_counts[key] = [0] * (len(self.buckets) + 1)
        for i, b in enumerate(self.buckets):
            if duration_ms <= b:
                self.bucket_counts[key][i] += 1
        self.bucket_counts[key][-1] += 1

    def format_prometheus(self) -> str:
        lines = []
        for key, count in self.request_count.items():
            lines.append(f'http_requests_total{{service="{self.service_name}",endpoint="{key}"}} {count}')
        for key, count in self.error_count.items():
            lines.append(f'http_errors_total{{service="{self.service_name}",endpoint="{key}"}} {count}')
        for key, buckets in self.bucket_counts.items():
            for i, b in enumerate(self.buckets):
                lines.append(f'http_request_duration_ms_bucket{{service="{self.service_name}",endpoint="{key}",le="{b}"}} {buckets[i]}')
            lines.append(f'http_request_duration_ms_bucket{{service="{self.service_name}",endpoint="{key}",le="+Inf"}} {buckets[-1]}')
            lines.append(f'http_request_duration_ms_sum{{service="{self.service_name}",endpoint="{key}"}} {self.latency_sum[key]:.3f}')
            lines.append(f'http_request_duration_ms_count{{service="{self.service_name}",endpoint="{key}"}} {self.latency_count[key]}')
        return "\n".join(lines)


class MetricsMiddleware(BaseHTTPMiddleware):
    def __init__(self, app, metrics: PrometheusMetrics):
        super().__init__(app)
        self.metrics = metrics

    async def dispatch(self, request: Request, call_next):
        start = time.time()
        response = await call_next(request)
        duration_ms = (time.time() - start) * 1000
        self.metrics.record(request.method, request.url.path, response.status_code, duration_ms)
        return response


class CircuitState(str, Enum):
    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half_open"


@dataclass
class CircuitBreaker:
    name: str
    max_failures: int = 5
    timeout: float = 30.0
    _failures: int = 0
    _state: CircuitState = CircuitState.CLOSED
    _last_failure: float = 0

    async def execute(self, fn: Callable, *args, **kwargs):
        if self._state == CircuitState.OPEN:
            if time.time() - self._last_failure > self.timeout:
                self._state = CircuitState.HALF_OPEN
            else:
                raise Exception(f"Circuit breaker {self.name} is open")

        try:
            if asyncio.iscoroutinefunction(fn):
                result = await fn(*args, **kwargs)
            else:
                result = fn(*args, **kwargs)
            self._failures = 0
            self._state = CircuitState.CLOSED
            return result
        except Exception as e:
            self._failures += 1
            self._last_failure = time.time()
            if self._failures >= self.max_failures:
                self._state = CircuitState.OPEN
            raise

    @property
    def state(self) -> str:
        return self._state.value


async def retry_with_backoff(fn: Callable, max_retries: int = 3, base_delay: float = 1.0, *args, **kwargs):
    last_error = None
    for i in range(max_retries + 1):
        try:
            if asyncio.iscoroutinefunction(fn):
                return await fn(*args, **kwargs)
            return fn(*args, **kwargs)
        except Exception as e:
            last_error = e
            if i < max_retries:
                delay = base_delay * (2 ** i) + (time.time() % 1) * base_delay
                await asyncio.sleep(delay)
    raise last_error


class StructuredLogger:
    def __init__(self, service: str):
        self.service = service
        self._logger = logging.getLogger(service)
        if not self._logger.handlers:
            handler = logging.StreamHandler()
            handler.setFormatter(logging.Formatter('%(message)s'))
            self._logger.addHandler(handler)
            self._logger.setLevel(logging.DEBUG)

    def _log(self, level: str, msg: str, **fields):
        entry = {
            "timestamp": time.strftime("%Y-%m-%dT%H:%M:%S.000Z", time.gmtime()),
            "level": level,
            "service": self.service,
            "message": msg,
            **fields,
        }
        self._logger.info(json.dumps(entry))

    def info(self, msg: str, **fields):
        self._log("info", msg, **fields)

    def error(self, msg: str, **fields):
        self._log("error", msg, **fields)

    def warn(self, msg: str, **fields):
        self._log("warn", msg, **fields)

    def debug(self, msg: str, **fields):
        self._log("debug", msg, **fields)


async def validate_jwt_token(keycloak_url: str, token: str) -> tuple[bool, str]:
    if not keycloak_url or not token:
        return False, ""
    try:
        import httpx
        async with httpx.AsyncClient(timeout=5) as client:
            resp = await client.post(
                f"{keycloak_url}/auth/validate",
                headers={"Authorization": f"Bearer {token}"},
            )
            if resp.status_code == 200:
                data = resp.json()
                return data.get("valid", False), data.get("user_id", "")
    except Exception:
        pass
    return False, ""


class JWTAuthMiddleware(BaseHTTPMiddleware):
    SKIP_PATHS = {"/health", "/metrics", "/prometheus", "/ready", "/live", "/docs", "/openapi.json"}

    def __init__(self, app, keycloak_url: str = ""):
        super().__init__(app)
        self.keycloak_url = keycloak_url

    async def dispatch(self, request: Request, call_next):
        if request.url.path in self.SKIP_PATHS:
            return await call_next(request)

        auth = request.headers.get("Authorization", "")
        if auth.startswith("Bearer ") and self.keycloak_url:
            token = auth[7:]
            valid, user_id = await validate_jwt_token(self.keycloak_url, token)
            if valid:
                request.state.user_id = user_id

        return await call_next(request)


def readiness_check(checks: dict[str, Callable[[], bool]]):
    async def handler():
        results = {}
        all_ready = True
        for name, check in checks.items():
            try:
                ok = check()
                results[name] = ok
                if not ok:
                    all_ready = False
            except Exception:
                results[name] = False
                all_ready = False
        return {"ready": all_ready, "checks": results}
    return handler

File: services/go/shared/middleware.go
package shared

import (
	"encoding/json"
	"fmt"
	"math"
	"math/rand"
	"net/http"
	"sync"
	"sync/atomic"
	"time"
)

type CorrelationIDKey struct{}

func TracingMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		corrID := r.Header.Get("X-Correlation-ID")
		if corrID == "" {
			corrID = fmt.Sprintf("corr-%d-%d", time.Now().UnixNano(), rand.Intn(10000))
		}
		traceID := r.Header.Get("X-Trace-ID")
		if traceID == "" {
			traceID = fmt.Sprintf("trace-%d", time.Now().UnixNano())
		}
		spanID := fmt.Sprintf("span-%d", time.Now().UnixNano())

		w.Header().Set("X-Correlation-ID", corrID)
		w.Header().Set("X-Trace-ID", traceID)
		w.Header().Set("X-Span-ID", spanID)

		r.Header.Set("X-Correlation-ID", corrID)
		r.Header.Set("X-Trace-ID", traceID)
		r.Header.Set("X-Span-ID", spanID)

		next.ServeHTTP(w, r)
	})
}

type PrometheusMetrics struct {
	mu             sync.Mutex
	requestCount   map[string]*atomic.Int64
	errorCount     map[string]*atomic.Int64
	latencyBuckets map[string]*LatencyHistogram
	serviceName    string
}

type LatencyHistogram struct {
	buckets []float64
	counts  []atomic.Int64
	sum     atomic.Int64
	count   atomic.Int64
}

func NewPrometheusMetrics(serviceName string) *PrometheusMetrics {
	return &PrometheusMetrics{
		requestCount:   make(map[string]*atomic.Int64),
		errorCount:     make(map[string]*atomic.Int64),
		latencyBuckets: make(map[string]*LatencyHistogram),
		serviceName:    serviceName,
	}
}

func (pm *PrometheusMetrics) getCounter(m map[string]*atomic.Int64, key string) *atomic.Int64 {
	pm.mu.Lock()
	defer pm.mu.Unlock()
	if c, ok := m[key]; ok {
		return c
	}
	c := &atomic.Int64{}
	m[key] = c
	return c
}

func (pm *PrometheusMetrics) getHistogram(key string) *LatencyHistogram {
	pm.mu.Lock()
	defer pm.mu.Unlock()
	if h, ok := pm.latencyBuckets[key]; ok {
		return h
	}
	h := &LatencyHistogram{
		buckets: []float64{5, 10, 25, 50, 100, 250, 500, 1000, 2500, 5000, 10000},
		counts:  make([]atomic.Int64, 12),
	}
	pm.latencyBuckets[key] = h
	return h
}

func (pm *PrometheusMetrics) RecordRequest(method, path string, statusCode int, durationMs float64) {
	key := fmt.Sprintf("%s_%s", method, path)
	pm.getCounter(pm.requestCount, key).Add(1)
	if statusCode >= 400 {
		pm.getCounter(pm.errorCount, key).Add(1)
	}
	h := pm.getHistogram(key)
	h.sum.Add(int64(durationMs * 1000))
	h.count.Add(1)
	for i, b := range h.buckets {
		if durationMs <= b {
			h.counts[i].Add(1)
		}
	}
	h.counts[len(h.buckets)].Add(1)
}

func (pm *PrometheusMetrics) Handler() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "text/plain; version=0.0.4")
		pm.mu.Lock()
		defer pm.mu.Unlock()

		for key, count := range pm.requestCount {
			fmt.Fprintf(w, "http_requests_total{service=\"%s\",endpoint=\"%s\"} %d\n", pm.serviceName, key, count.Load())
		}
		for key, count := range pm.errorCount {
			fmt.Fprintf(w, "http_errors_total{service=\"%s\",endpoint=\"%s\"} %d\n", pm.serviceName, key, count.Load())
		}
		for key, h := range pm.latencyBuckets {
			for i, b := range h.buckets {
				fmt.Fprintf(w, "http_request_duration_ms_bucket{service=\"%s\",endpoint=\"%s\",le=\"%.0f\"} %d\n", pm.serviceName, key, b, h.counts[i].Load())
			}
			fmt.Fprintf(w, "http_request_duration_ms_bucket{service=\"%s\",endpoint=\"%s\",le=\"+Inf\"} %d\n", pm.serviceName, key, h.counts[len(h.buckets)].Load())
			fmt.Fprintf(w, "http_request_duration_ms_sum{service=\"%s\",endpoint=\"%s\"} %.3f\n", pm.serviceName, key, float64(h.sum.Load())/1000)
			fmt.Fprintf(w, "http_request_duration_ms_count{service=\"%s\",endpoint=\"%s\"} %d\n", pm.serviceName, key, h.count.Load())
		}
	}
}

func MetricsMiddleware(metrics *PrometheusMetrics, next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()
		rw := &responseWriter{ResponseWriter: w, statusCode: 200}
		next.ServeHTTP(rw, r)
		durationMs := float64(time.Since(start).Microseconds()) / 1000
		metrics.RecordRequest(r.Method, r.URL.Path, rw.statusCode, durationMs)
	})
}

type responseWriter struct {
	http.ResponseWriter
	statusCode int
}

func (rw *responseWriter) WriteHeader(code int) {
	rw.statusCode = code
	rw.ResponseWriter.WriteHeader(code)
}

func JWTAuthMiddleware(keycloakURL string, next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path == "/health" || r.URL.Path == "/metrics" || r.URL.Path == "/prometheus" || r.URL.Path == "/ready" || r.URL.Path == "/live" {
			next.ServeHTTP(w, r)
			return
		}

		authHeader := r.Header.Get("Authorization")
		if authHeader == "" {
			next.ServeHTTP(w, r)
			return
		}

		if len(authHeader) > 7 && authHeader[:7] == "Bearer " {
			token := authHeader[7:]
			valid, userID := validateToken(keycloakURL, token)
			if valid {
				r.Header.Set("X-User-ID", userID)
			}
		}

		next.ServeHTTP(w, r)
	})
}

func validateToken(keycloakURL, token string) (bool, string) {
	if keycloakURL == "" || token == "" {
		return false, ""
	}
	client := &http.Client{Timeout: 5 * time.Second}
	req, err := http.NewRequest("POST", keycloakURL+"/auth/validate", nil)
	if err != nil {
		return false, ""
	}
	req.Header.Set("Authorization", "Bearer "+token)
	resp, err := client.Do(req)
	if err != nil {
		return false, ""
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return false, ""
	}
	var result struct {
		Valid  bool   `json:"valid"`
		UserID string `json:"user_id"`
	}
	json.NewDecoder(resp.Body).Decode(&result)
	return result.Valid, result.UserID
}

type CircuitBreaker struct {
	name          string
	maxFailures   int
	timeout       time.Duration
	failures      atomic.Int64
	state         atomic.Int32
	lastFailureAt atomic.Int64
}

const (
	StateClosed   int32 = 0
	StateOpen     int32 = 1
	StateHalfOpen int32 = 2
)

func NewCircuitBreaker(name string, maxFailures int, timeout time.Duration) *CircuitBreaker {
	return &CircuitBreaker{
		name:        name,
		maxFailures: maxFailures,
		timeout:     timeout,
	}
}

func (cb *CircuitBreaker) Execute(fn func() error) error {
	state := cb.state.Load()

	if state == StateOpen {
		lastFail := time.Unix(0, cb.lastFailureAt.Load())
		if time.Since(lastFail) > cb.timeout {
			cb.state.CompareAndSwap(StateOpen, StateHalfOpen)
		} else {
			return fmt.Errorf("circuit breaker %s is open", cb.name)
		}
	}

	err := fn()
	if err != nil {
		failures := cb.failures.Add(1)
		cb.lastFailureAt.Store(time.Now().UnixNano())
		if failures >= int64(cb.maxFailures) {
			cb.state.Store(StateOpen)
		}
		return err
	}

	cb.failures.Store(0)
	cb.state.Store(StateClosed)
	return nil
}

func (cb *CircuitBreaker) State() string {
	switch cb.state.Load() {
	case StateClosed:
		return "closed"
	case StateOpen:
		return "open"
	case StateHalfOpen:
		return "half-open"
	}
	return "unknown"
}

func RetryWithBackoff(maxRetries int, baseDelay time.Duration, fn func() error) error {
	var lastErr error
	for i := 0; i <= maxRetries; i++ {
		lastErr = fn()
		if lastErr == nil {
			return nil
		}
		if i < maxRetries {
			delay := time.Duration(float64(baseDelay) * math.Pow(2, float64(i)))
			jitter := time.Duration(rand.Int63n(int64(delay / 2)))
			time.Sleep(delay + jitter)
		}
	}
	return fmt.Errorf("after %d retries: %w", maxRetries, lastErr)
}

type StructuredLogger struct {
	service string
}

func NewLogger(service string) *StructuredLogger {
	return &StructuredLogger{service: service}
}

func (l *StructuredLogger) log(level, msg string, fields map[string]interface{}) {
	entry := map[string]interface{}{
		"timestamp": time.Now().UTC().Format(time.RFC3339Nano),
		"level":     level,
		"service":   l.service,
		"message":   msg,
	}
	for k, v := range fields {
		entry[k] = v
	}
	data, _ := json.Marshal(entry)
	fmt.Println(string(data))
}

func (l *StructuredLogger) Info(msg string, fields map[string]interface{}) {
	l.log("info", msg, fields)
}

func (l *StructuredLogger) Error(msg string, fields map[string]interface{}) {
	l.log("error", msg, fields)
}

func (l *StructuredLogger) Warn(msg string, fields map[string]interface{}) {
	l.log("warn", msg, fields)
}

func (l *StructuredLogger) Debug(msg string, fields map[string]interface{}) {
	l.log("debug", msg, fields)
}

func ReadinessHandler(checks map[string]func() bool) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		allReady := true
		results := make(map[string]bool)
		for name, check := range checks {
			ok := check()
			results[name] = ok
			if !ok {
				allReady = false
			}
		}
		w.Header().Set("Content-Type", "application/json")
		if !allReady {
			w.WriteHeader(http.StatusServiceUnavailable)
		}
		json.NewEncoder(w).Encode(map[string]interface{}{
			"ready":  allReady,
			"checks": results,
		})
	}
}

func LivenessHandler() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]interface{}{
			"alive":     true,
			"timestamp": time.Now().UTC().Format(time.RFC3339),
		})
	}
}

type HTTPClient struct {
	client  *http.Client
	breaker *CircuitBreaker
	logger  *StructuredLogger
}

func NewHTTPClient(serviceName string, timeout time.Duration, maxFailures int) *HTTPClient {
	return &HTTPClient{
		client: &http.Client{
			Timeout: timeout,
			Transport: &http.Transport{
				MaxIdleConns:        100,
				MaxIdleConnsPerHost: 20,
				IdleConnTimeout:     90 * time.Second,
			},
		},
		breaker: NewCircuitBreaker(serviceName, maxFailures, 30*time.Second),
		logger:  NewLogger(serviceName),
	}
}

func (hc *HTTPClient) Do(req *http.Request) (*http.Response, error) {
	var resp *http.Response
	err := hc.breaker.Execute(func() error {
		var err error
		resp, err = hc.client.Do(req)
		if err != nil {
			return err
		}
		if resp.StatusCode >= 500 {
			return fmt.Errorf("server error: %d", resp.StatusCode)
		}
		return nil
	})
	return resp, err
}

func (hc *HTTPClient) CircuitState() string {
	return hc.breaker.State()
}
